lib/routes/routeUtapi.js

```diff
@@ -0,0 +1,76 @@
+const url = require('url');
+const http = require('http');
+const { errors } = require('arsenal');
+const { responseJSONBody } = require('arsenal').s3routes.routesUtils;
+const { config } = require('../Config');
+
+function _decodeURI(uri) {
+    // do the same decoding as in S3 server
+    return decodeURIComponent(uri.replace(/\+/g, ' '));
+}
+
+function _createUtapiRequestOptions(req) {
+    const parsedUrl = url.parse(req.url, true);
+    const reqPath = _decodeURI(parsedUrl.pathname);
+    // path will begin with /_/utapi
+    const utapiPath = `${reqPath.substring(8)}${parsedUrl.search}`;
+    const utapiPort = config.utapi.port || 8100;
+    const utapiHost = config.utapi.host || 'localhost';
+
+    return {
+        path: utapiPath,
+        port: utapiPort,
+        host: utapiHost,
+        method: 'POST',
+        headers: {
+            'content-type': 'application/json',
+            'cache-control': 'no-cache',
+        },
+        rejectUnauthorized: false,
+    };
+}
+
+function _respond(response, payload, log) {
+    const body = typeof payload === 'object' ?
+        JSON.stringify(payload) : payload;
+    const httpHeaders = {
+        'x-amz-id-2': log.getSerializedUids(),
+        'x-amz-request-id': log.getSerializedUids(),
+        'content-type': 'application/json',
+        'content-length': Buffer.byteLength(body),
+    };
+    response.writeHead(200, httpHeaders);
+    response.end(body, 'utf8');
+}
+
+
+function routeUtapi(clientIP, request, response, log) {
+    log.debug('route request', { method: 'routeUtapi' });
+    let reqBody = '';
+    request.on('data', chunk => {
+        reqBody += chunk.toString();
```

Comment:
It's been my understanding that pushing to an array and then calling `.join` (as you've done below) has better performance than concatenating in this manner. Maybe that's inaccurate though—I actually couldn't find a definite source for that idea 😺.

Reply:
In this case, I tried the `.join` way of creating the `reqBody`, but when it was written to the new request, the format was messed up. This was the only way I was able to keep the JSON properly formatted on the passthrough to Utapi :)